text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,459 @@
+
+```rust
+use std::{sync::{Mutex, Lazy}, collections::HashMap};
+
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>> = Lazy::new(|| {
+    let mut m = HashMap::new();
+    m.insert(13, "Spica".to_string());
+    m.insert(74, "Hoyten".to_string());
+    Mutex::new(m)
+});
+```
+
+Moreover, once `#[thread_local]` attribute is stable (which is not on the road to stabilization yet), `Lazy` will supplant `std::thread_local!` as well:
+
+```rust
+use std::cell::{RefCell, Lazy};
+
+#[thread_local]
+pub static FOO: Lazy<RefCell<u32>> = Lazy::new(|| RefCell::new(1));
+```
+
+Unlike `lazy_static!`, `Lazy` can be used used for locals:
+
+```rust
+use std::cell::Lazy;
+
+fn main() {
+    let ctx = vec![1, 2, 3];
+    let thunk = Lazy::new(|| {
+        ctx.iter().sum::<i32>()
+    });
+    assert_eq!(*thunk, 6);
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The proposed API is directly copied from [`once_cell`] crate.
+
+Altogether, this RFC proposes to add four types:
+
+* `std::cell::OnceCell`, `std::cell::Lazy`
+* `std::sync::OnceCell`, `std::sync::Lazy`
+
+`OnceCell` is an important core primitive.
+`Lazy` can be stabilized separately from `OnceCell`, or it can be omitted from the standard library altogether.
+However, it provides significantly nicer ergonomics for the common use-case of static lazy values.
+
+Non thread-safe flavor is implemented by storing an `UnsafeCell<Option<T>>`:
+
+```rust
+pub struct OnceCell<T> {
+    // Invariant: written to at most once.
+    inner: UnsafeCell<Option<T>>,
+}
+```
+
+The implementation is mostly straightforward.
+The only tricky bit is that reentrant initialization should be explicitly forbidden.
+That is, the following program panics:
+
+```rust
+let x: OnceCell<Box<i32>> = OnceCell::new();
+let dangling_ref: Cell<Option<&i32>> = Cell::new(None);
+x.get_or_init(|| {
+    let r = x.get_or_init(|| Box::new(92));
+    dangling_ref.set(Some(r));
+    Box::new(62)
+});
+println!("would be use after free: {:?}", dangling_ref.get().unwrap());
+```
+
+Non thread-safe flavor can be added to `core` as well.
+
+The thread-safe variant is implemented similarly to `std::sync::Once`.
+Crucially, it has support for blocking: if many threads call `get_or_init` concurrently, only one will be able to execute the closure, while all other threads will block.
+For this reason, most of `std::sync::OnceCell` API can not be provided in `core`.
+In the `sync` case, reliably panicking on re-entrant initialization is not trivial.
+For this reason, the implementaion would simply deadlock, with a note that a deadlock might be elevated to panic in the future.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This is a moderately large addition to stdlib, there's a chance we do something wrong.
+  This can be mitigated by piece-wise stabilization (in particular, `Lazy` convenience types are optional) and the fact that API is battle-tested via `once_cell` crate.
```

Comment:
nit: maybe prefer "widely used", "heavily audited", or some other more concrete and less aggressive description than "battle-tested"?